### PR TITLE
feat: Adds a new raw file metadata storage for clients

### PR DESCRIPTION
**Pull request by kommendorkapten**
This came up during a discussion related to new features for the sigstore TUF client, as there is a desire to make implementations in different languages sharing the target and metadata cache.

Release Notes: Possibility to store metadata files as raw JSON files to allow for interoperability with other TUF implementations.

**Types of changes**:
- [  ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [  ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**Description of the changes being introduced by the pull request**:
This pull request introduces a new implementation of the client's `LocalStore` interface. This storage stores the metadata files are regular files in a provided directory. This would improve compatibility to share metadata cache between different language implementations.

**Please verify and check that the pull request fulfills the following requirements**:

- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: client/filejsonstore/filejsonstore_test.go
Comment (joshuagl) [inline client/filejsonstore/filejsonstore_test.go:18]: Nit: typo
```suggestion
// Hook up gocheck into the "go test" runner
```

**Review by joshuagl — APPROVED** [review]
Thanks, this LGTM. 

It would be great to get a review from a maintainer who is more fluent in Go than I am.

**Review by asraa — COMMENTED** [review]
The change looks good to me as a start, but I wonder a few things and would love feedback:

 Do we need a lock mechanism, esp for SetMeta()? If there are more than one client accessing the store.

**Comment by kommendorkapten** [comment]
Yeah, let me add version that's safe for concurrent use!

**Comment by kommendorkapten** [comment]
I added a new wrapper, 'ConcurrentLocalStore' which can be used to protect any `LocalStore` (e.g. the in memory one and the raw JSON one) for concurrent access. 
This would let the client decide what use depending on their access pattern.

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:79]: This line will never run. Let's check the error before we check `ok`.

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:41]: Perhaps we should check permissions on the directory and the files inside (in `GetMeta`) if it already exists, similar to how SSH validates keys on disk.

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:68]:
```suggestion
```

File: client/filejsonstore/filejsonstore_test.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore_test.go:91]:
```suggestion
	// Delete all items
```

File: client/filejsonstore/filejsonstore_test.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore_test.go:116]: Can we check the returned error?

File: client/filejsonstore/filejsonstore_test.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore_test.go:139]:
```suggestion
```

File: client/filejsonstore/filejsonstore_test.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore_test.go:172]:
```suggestion
	tmp := c.MkDir()
	p := filepath.Join(tmp, "tuf_raw.db")
	localStore, err := NewFileJSONStore(p)
```

File: client/filejsonstore/filejsonstore_test.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore_test.go:16]: Can we name these tests consistently? (`RawJSONStoreSuite -> FileJSONStoreSuite`)

File: client/concurrentlocalstore.go
Comment (kommendorkapten) [reply client/concurrentlocalstore.go:10]: No particular reason except it's flexible. Any `LocalStore` can be protected with this construct, and the existing ones did not perform any concurrency control, so I figured it can be solved for them all with this. And It _can_ simplify development as the locking is managed by a different layer. I'm fine to refactor if it's desired.

There can definitely be some scenarios that does not require concurrency control, but I hardly think there is a measurable performance benefit from omitting it :)

Comment (ethan-lowman-dd) [reply client/concurrentlocalstore.go:10]: Yeah, intuitively, I would think that if no concurrency control is actually needed, there would be zero lock contention, and thus minimal overhead for the default safety.

File: client/filejsonstore/filejsonstore_test.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore_test.go:171]: The reason here for being upfront with the type (`client.LocalStore`) is to make sure the returned type is matching this interface, so I rather not change it.

File: util/util.go
Comment (kommendorkapten) [reply util/util.go:336]: 👍

File: client/filejsonstore/filejsonstore_test.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore_test.go:116]: 👍

File: client/filejsonstore/filejsonstore_test.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore_test.go:16]: 👍

File: client/filejsonstore/filejsonstore.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore.go:44]: Yes, you are correct. I have a faint memory of this being required in an earlier version, but now as you said it's not needed.

File: client/filejsonstore/filejsonstore.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore.go:78]: Good catch!

File: client/filejsonstore/filejsonstore.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore.go:41]: Sure, I'll add checks to make sure only `user` (owner) have write permission to the directory and any files in it.

**Comment by kommendorkapten** [comment]
Windows tests are failing, I will look into this tomorrow.

**Comment by kommendorkapten** [comment]
The failing tests are due to filesystem permission, which on Windows is very different from UNIX-like operating system.
I found this post on the topic: https://medium.com/@MichalPristas/go-and-file-perms-on-windows-3c944d55dd44
The summary is that in windows you can only create a file with either:
* 0666 (Read/writeable by all users)
* 0444 (Readable by all users)

And this seems to be correct according the source code: https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L307
The provided permission is only used to determine if a file is intended to be create as read-only. 
The same is valid for `Mkdir`, the permission mode bits are ignored on Windows: https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L518
And for reference the `Chmod` call: https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L646 (only the read-only flag is manipulated).

It seems that there are some ACL concept in Windows that we can rely on. But I'm not qualified to speak if the normal permission modes set (all-read or all-write) actually are system-wide or if the effective permission applies to the user only. I found this ACL implementation in go: https://github.com/hectane/go-acl
It haven't had any updates in three years, and I would not be comfortable relying on this, as I lack understanding on Windows filesystem access semantics.

As this feature of verifying the filesystem permission bits, is new for this implementation (LevelDB implementation does not care), nor is there any guidance for implementors of `LocalStore` to consider this my suggestion is _to skip file permissions for Windows_.

Please advise on how to proceed.

File: client/filejsonstore/filejsonstore.go
Comment (patflynn) [inline client/filejsonstore/filejsonstore.go:45]: (s)afe

**Comment by kommendorkapten** [comment]
Any updates on how to proceed with this (file permission on Windows host)? My comments and proposal are in the comment above: https://github.com/theupdateframework/go-tuf/pull/347#issuecomment-1211615181
cc @trishankatdatadog @ethan-lowman-dd

**Comment by trishankatdatadog** [comment]
> Any updates on how to proceed with this (file permission on Windows host)? My comments and proposal are in the comment above: [#347 (comment)](https://github.com/theupdateframework/go-tuf/pull/347#issuecomment-1211615181) cc @trishankatdatadog @ethan-lowman-dd

I'm going to make a point of looking at this tonight.

**Comment by trishankatdatadog** [comment]
> Please advise on how to proceed.

Sorry for the late reply.

Let's take a look at the python-tuf [reference implementation](https://github.com/theupdateframework/python-tuf/blob/7210487281ee24d14ebfcdc7c1543d2751113c5f/tuf/ngclient/updater.py): there, the permissions of the metadata/targets cache directories is not checked, and metadata/targets files appear to be RW-able by the user but not others (using both `tempfile.NamedTemporaryFile()` and `open()`).

I like your idea for checking the permissions on both the directories and files, but technically, this is not a requirement in the TUF specification. At least the directory permissions can probably be left to the application calling go-tuf, but the files inside them are trickier. If the golang stdlib doesn't give control beyond either RW or RO by all, then there is not much we can do for now.

Is not supporting Windows for now an option? That seems the safest to me.

**Comment by trishankatdatadog** [comment]
@kommendorkapten do you need another review? If so, I recommend pinging those who have reviewed before. Thanks!

**Comment by kommendorkapten** [comment]
@ethan-lowman-dd would you mind taking another look? Addressed your concerns and added filesystem permission check for the metadata files/directories on UNIX-like systems (see discussion above on why Windows is exempted).

**Comment by trishankatdatadog** [comment]
@kommendorkapten if there isn't already a GitHub issue to track the Windows file permissions issue, would you mind making one for it? We should track it. Thanks!

**Comment by kommendorkapten** [comment]
@trishankatdatadog Of course! Created this> https://github.com/theupdateframework/go-tuf/issues/360

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:23]: Can we make these not sentinel errors and use `fmt.Sprintf` with context for debugging instead? See `perm.go` for more discussion.

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:50]: Can we initialize as a pointer? Now that there is a mutex field, the only valid way to use this type is via a pointer type, since otherwise the mutex is copied.

```suggestion
	f := &FileJSONStore{
```

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:53]: This is unnecessary
```suggestion
```

File: client/filejsonstore/filejsonstore.go
Comment (ethan-lowman-dd) [inline client/filejsonstore/filejsonstore.go:94]: nit: can we put `p` closer to where it's used?

Also, is `filepath.FromSlash` necessary, given that `filepath.Join` already uses an OS-specific filepath separator?

File: internal/fsutil/perm.go
Comment (kommendorkapten) [reply internal/fsutil/perm.go]: Ah, nice.

File: client/filejsonstore/filejsonstore.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore.go:50]: Yeah. The result are the same though as the pointer escapes this function, and all methods receives the target by pointer. But making it a pointer in the first place could be more clear.

File: client/filejsonstore/filejsonstore.go
Comment (kommendorkapten) [reply client/filejsonstore/filejsonstore.go:94]: The issue is that we don't know what's inside `f.baseDir`, but that could be fixed in `NewFileJSONStore`. But as I read through the code of `Join` I found that prior to returning the value, `Clean` is called, and `Clean` calls `FromSlash`: https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/path/filepath/path.go;l=89

So yes, this `FromSlash` is redundant.

**Comment by kommendorkapten** [comment]
Good feedback @ethan-lowman-dd, all your comments should be addressed now 👍

**Review by ethan-lowman-dd — DISMISSED** [review]

**Comment by ethan-lowman-dd** [comment]
@asraa or @joshuagl Could you please provide a followup review when you get the chance?

**Comment by trishankatdatadog** [comment]
> Good feedback @ethan-lowman-dd, all your comments should be addressed now 👍

Nice work, thanks Ethan

**Review by joshuagl — APPROVED** [review]
Lovely work here @kommendorkapten and @ethan-lowman-dd , thank you!

**Comment by trishankatdatadog** [comment]
@kommendorkapten would you please rebase and resolve conflicts? thx!

**Comment by kommendorkapten** [comment]
Was off on PTO, so didn't see you message @trishankatdatadog until now. PR is updated now.

**Comment by trishankatdatadog** [comment]
> Was off on PTO, so didn't see you message @trishankatdatadog until now. PR is updated now.

Thanks! Would you pls fix the linting errors?

**Comment by kommendorkapten** [comment]
> Thanks! Would you pls fix the linting errors?

Sorry, was not aware of them, fixed now.

**Review by ethan-lowman-dd — APPROVED** [review]

**Review by trishankatdatadog — APPROVED** [review]

**Pull request merged**
